pipelines/build/common/config_regeneration.groovy

```diff
@@ -0,0 +1,433 @@
+        CONFIGURE_ARGS: getConfigureArgs(platformConfig, variant),
+        OVERRIDE_FILE_NAME_VERSION: "",
+        ADDITIONAL_FILE_NAME_TAG: platformConfig.additionalFileNameTag as String,
+        JDK_BOOT_VERSION: platformConfig.bootJDK as String, 
+        RELEASE: false,
+        PUBLISH_NAME: "",
+        ADOPT_BUILD_NUMBER: "",
+        ENABLE_TESTS: true,
+        CLEAN_WORKSPACE: true
+      )
+    } catch (Exception e) {
+      // Catch invalid configurations
+      context.println "[ERROR] Failed to create IndividualBuildConfig for platformConfig: ${platformConfig}.\nError: ${e}"
+      currentBuild.result = "FAILURE"
+    }
+  }
+
+  /**
+  * Checks if the parameter is a map
+  * @param possibleMap
+  */
+  static def isMap(possibleMap) {
+    return Map.class.isInstance(possibleMap)
+  }
+
+  /**
+  * Generates a job from template at `create_job_from_template.groovy`. This is what creates the job dsl and "regenerates" the job.
+  * @param jobName
+  * @param jobFolder
+  * @param config
+  */
+  def createJob(jobName, jobFolder, IndividualBuildConfig config) {
+    Map<String, ?> params = config.toMap().clone() as Map
+    params.put("JOB_NAME", jobName)
+    params.put("JOB_FOLDER", jobFolder)
+
+    params.put("GIT_URI", "https://github.com/AdoptOpenJDK/openjdk-build.git")
+    params.put("GIT_BRANCH", "master") 
+
+    params.put("BUILD_CONFIG", config.toJson())
+
+    def create = context.jobDsl targets: "pipelines/build/common/create_job_from_template.groovy", ignoreExisting: false, additionalParameters: params
+
+    return create
+  }
+
+  /**
+  * Queries the Jenkins API. Used to get the pipeline and downstream job details.
+  * @param query
+  */
+  def queryJenkinsAPI(String query) {
+    try {
+      def parser = new JsonSlurper()
+      def get = new URL(query).openConnection()
+      def response = parser.parseText(get.getInputStream().getText())
+      return response
+
+    } catch (Exception e) {
+      // Failed to connect to jenkins api or a parsing error occured
+      context.println "[ERROR] Failure on jenkins api connection or parsing.\nError: ${e}"
+      currentBuild.result = "FAILURE"
+    } 
+  }
+
+  /**
+  * Main function. Ran from regeneration_pipeline.groovy, this will be what jenkins will run first. 
+  */ 
+  @SuppressWarnings("unused")
+  def regenerate() {
+    /*
+    * Stage: Check that the pipeline isn't in inprogress or queued up. Once clear, run the regeneration job
+    */
+    context.stage("Check $javaVersion pipeline status") {
+
+      // Get all pipelines
+      def getPipelines = queryJenkinsAPI("https://ci.adoptopenjdk.net/job/build-scripts/api/json?tree=jobs[name]&pretty=true&depth1")
+
+      // Parse api response to only extract the relevant pipeline
+      getPipelines.jobs.name.each{ pipeline -> 
+        if (pipeline.contains("pipeline") && pipeline.contains(javaVersion)) {
+          Integer sleepTime = 900
+          Boolean inProgress = true
+
+          while (inProgress) {
```

Comment:
This logic of individually looking at builds 1 by 1 and waiting for it to end does not seem like it will work as builds can (and probably will) start while you are waiting for others to end. Would it not be better to regenerate pipelines 1 by 1 if they are not running rather than trying to wait for a time when all of them are not running?

Reply:
> rather than trying to wait for a time when all of them are not running?
> 

That is what it is currently doing. The idea is have multiple regeneration jobs for each java version. Each one only checks if its corresponding pipeline is running
- https://ci.adoptopenjdk.net/view/work%20in%20progress/job/morgan_pipeline_generator_jdk8u/3/console
- https://ci.adoptopenjdk.net/view/work%20in%20progress/job/morgan_pipeline_generator_jdk11u/3/console
- https://ci.adoptopenjdk.net/view/work%20in%20progress/job/morgan_pipeline_generator_jdk14u/5/console
- https://ci.adoptopenjdk.net/view/work%20in%20progress/job/morgan_pipeline_generator_jdk/3/console